### ISSUE TITLE
EPERM: operation not permitted, unlink '...node_modules\prisma\query_engine-windows.dll.node'

### ISSUE THREAD

**Issue by RedGlow** [prisma/prisma#9184]
### Bug description

Following the [Getting Started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) guide, when it gets to the point of running the "migrate dev" command, the script tries to install some packages through npm (don't know which ones) and this causes the error in title.

### How to reproduce

1. Follow the [Getting started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) prisma guide up to and included the "npx prisma migrate dev --name init" command.
2. Get the following error:

```
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": PostgreSQL database "testprisma", schema "public" at "localhost:5432"

PostgreSQL database testprisma created at localhost:5432

The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20210909080511_init/
    └─ migration.sql

Your database is now in sync with your schema.

Running generate... (Use --skip-generate to skip the generators)
npm WARN hello-prisma@1.0.0 No description
npm WARN hello-prisma@1.0.0 No repository field.

npm ERR! code EPERM
npm ERR! syscall unlink
npm ERR! path C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node
npm ERR! errno -4048
npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'
npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!     errno: -4048,
npm ERR!     code: 'EPERM',
npm ERR!     syscall: 'unlink',
npm ERR!     path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node'
npm ERR!   },
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node',
npm ERR!   parent: 'hello-prisma'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\mbelletti\AppData\Roaming\npm-cache\_logs\2021-09-09T08_05_23_557Z-debug.log
Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.1
The batch file cannot be found.
```

The log file is the following:
[2021-09-09T08_05_23_557Z-debug.log](https://github.com/prisma/prisma/files/7134677/2021-09-09T08_05_23_557Z-debug.log)

After this, the node_modules\.bin directory no longer has the prisma executable, and the node_modules/prisma folder only contains the file that npm wasn't able to remove (query_engine-windows.dll.node).

### Expected behavior

The script finishes correctly.

### Prisma information

All the prisma configurations are the ones taken from the guide.

### Environment & setup

- OS: Windows 10
- Database: PostgreSQL 13
- Node.js version: 14.17.1 (using nvm)

### Prisma Version

```
Environment variables loaded from .env
prisma                  : 3.0.1
@prisma/client          : Not found
Current platform        : windows
Query Engine (Node-API) : libquery-engine 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 2452cc6313d52b8b9a96999ac0e974d0aedf88db
Studio                  : 0.423.0
```

(I had to relaunch npm install to get prisma back)

**Comment by pantharshit00** [comment]
The script tries to install `@prisma/client and `prisma` if they doesn't exists on your package.json file. 

I wasn't really able to reproduce this. Looks like the system you were using didn't give the migrate command the permission to save the Node API library that we download and use. 


Can you consistently reproduce this on your machine?

**Comment by RedGlow** [comment]
> The script tries to install `@prisma/client` and `prisma` if they doesn't exists on your package.json file.

This is interesting, because prisma is already present in my package.json, but from what I see the script is trying to re-install prisma too (I guess so because the package getting disrupted is prisma itself). Could be a version problem?

> Can you consistently reproduce this on your machine?
Yes, on this machine it happens consistently, and yesterday evening I had the same problem on another machine (but I should check it today, I tried just once).

As an additional information, after the script breaks, the node_modules/prisma directory only contains  query_engine-windows.dll.node. Could it be that the file is opened for some reason while the install script is running, and because of this it cannot be removed?

**Comment by pantharshit00** [comment]
>  Could it be that the file is opened for some reason while the install script is running, and because of this it cannot be removed?

Might be possible. Did you try completely removing `node_modules` once and trying to reproduce this again?

**Comment by RedGlow** [comment]
I've tried right now by removing the node_modules directory, re-installing the packages with `npm install` and re-running `npx prisma migrate dev --name init`, and got the same result:

```
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": PostgreSQL database "testprisma", schema "public" at "localhost:5432"

Already in sync, no schema change or pending migration was found.

Running generate... (Use --skip-generate to skip the generators)
npm WARN hello-prisma@1.0.0 No description
npm WARN hello-prisma@1.0.0 No repository field.

npm ERR! code EPERM
npm ERR! syscall unlink
npm ERR! path C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node
npm ERR! errno -4048
npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'
npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!     errno: -4048,
npm ERR!     code: 'EPERM',
npm ERR!     syscall: 'unlink',
npm ERR!     path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node'
npm ERR!   },
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node',
npm ERR!   parent: 'hello-prisma'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\mbelletti\AppData\Roaming\npm-cache\_logs\2021-09-10T07_27_58_812Z-debug.log
Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.1
The batch file cannot be found.
```

**Comment by janpio** [comment]
I can confirm this problem:

```
C:\Users\Jan\Documents\throwaway\sethRepro>npx prisma migrate dev
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": SQLite database "fancy.db" at "file:./fancy.db"

SQLite database fancy.db created at file:./fancy.db

√ Enter a name for the new migration: ... init
The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20210910203934_init/
    └─ migration.sql

Your database is now in sync with your schema.

Running generate... (Use --skip-generate to skip the generators)
npm WARN sethRepro@1.0.0 No description
npm WARN sethRepro@1.0.0 No repository field.

npm ERR! code EPERM
npm ERR! syscall unlink
npm ERR! path C:\Users\Jan\Documents\throwaway\sethRepro\node_modules\prisma\query_engine-windows.dll.node
npm ERR! errno -4048
npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\Jan\Documents\throwaway\sethRepro\node_modules\prisma\query_engine-windows.dll.node'
npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\Jan\Documents\throwaway\sethRepro\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\Jan\Documents\throwaway\sethRepro\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!     errno: -4048,
npm ERR!     code: 'EPERM',
npm ERR!     syscall: 'unlink',
npm ERR!     path: 'C:\\Users\\Jan\\Documents\\throwaway\\sethRepro\\node_modules\\prisma\\query_engine-windows.dll.node'
npm ERR!   },
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path: 'C:\\Users\\Jan\\Documents\\throwaway\\sethRepro\\node_modules\\prisma\\query_engine-windows.dll.node',
npm ERR!   parent: 'sethRepro'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Jan\AppData\Roaming\npm-cache\_logs\2021-09-10T20_39_44_431Z-debug.log
Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.2
The batch file cannot be found.
```

Here are my versions:

```
C:\Users\Jan\Documents\throwaway\sethRepro\prisma
λ node -v
v14.17.5

C:\Users\Jan\Documents\throwaway\sethRepro\prisma
λ npm -v
6.14.14
```

Interesting is what happened on `npx prisma -v`:
```
C:\Users\Jan\Documents\throwaway\sethRepro\prisma
λ npx prisma -v
Environment variables loaded from ..\.env
prisma                  : 3.0.2
@prisma/client          : Not found
Current platform        : windows
Query Engine (Node-API) : libquery-engine 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at ..\..\..\..\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at ..\..\..\..\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at ..\..\..\..\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at ..\..\..\..\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 2452cc6313d52b8b9a96999ac0e974d0aedf88db
Studio                  : 0.423.0
Error: EPERM: operation not permitted, unlink 'C:\Users\Jan\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\query_engine-windows.dll.node'
    at Object.unlinkSync (fs.js:1251:3)
    at fixWinEPERMSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:219:13)
    at rimrafSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:319:28)
    at C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:350:5
    at Array.forEach (<anonymous>)
    at rmkidsSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:349:26)
    at rmdirSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:342:7)
    at fixWinEPERMSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:217:5)
    at rimrafSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:319:28)
    at C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:350:5
```

When running `npx prisma -v` it downloaded/installed something on the spot I think, so it might indeed be that the `prisma` installation somehow got corrupted. I have that sometimes but never could reproduce or understand it.

That my engine is located in `AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\prisma-fmt-windows.exe` is also an indication that something is weird and this is not the actual `prisma` in my project, but a temporary one from `npx`.

**Comment by janpio** [comment]
(This seems to happen on the in-the-background `npm install @prisma/client` and then corrupts the local `prisma` installation in the project, causing it to use an instant one via `npx`)

**Comment by NordicBeaver** [comment]
Can confirm, just had the same error. 
Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.

1. Delete `node_modules`
2. Run `npm install`
3. Run `npm install @prisma/client`
4. Run `npx prisma generate`

**Comment by janpio** [comment]
Can you shared your node and npm version and how you installed it? I do not really think this is anything wrong with Prisma itself, but just a weird combination of different things in the setup leading to some broken behavior.

**Comment by NordicBeaver** [comment]
> 
> 
> Can you shared your node and npm version and how you installed it? I do not really think this is anything wrong with Prisma itself, but just a weird combination of different things in the setup leading to some broken behavior.

Sure. I was losing my mind over this, so I even reinstalled node. The versions are `node: v14.17.6`, `npm: 6.14.15` on `Windows 10.0.19043`.
I can get the error simply by creating empy folder, running `npm init` and `npm install prisma`, then adding a simple model and running `npx prisma generate`.

Btw, I just tried doing the same with the older prisma version (2.20.3) and I didn't get the error.

**Comment by ignHaxiel** [comment]
Hello, updating my npm to the latest solve this error for me.

"In my situation this helped:

Before proceeding to execute these commands close all VS Code instances.

clean cache with

 npm cache clean --force
install the latest version of npm globally as admin:

 npm install -g npm@latest --force
clean cache with

 npm cache clean --force
Try to install your component once again.

I hope this fix your issue. If not, you may temporarily disable your antivirus program and try again."

source : https://stackoverflow.com/questions/39293636/npm-err-error-eperm-operation-not-permitted-rename

**Comment by LewisTehMinerz** [comment]
Can replicate with yarn:
![](https://discord.coffee/rfLgNHy.png)

**Comment by janpio** [comment]
> The fix provided by @NordicBeaver works perfectly and can bypass the issue.

Just to confirm, with that you mean a manual `npm install @prisma/client` before you first run `npx prisma generate`, correct?

**Comment by LewisTehMinerz** [comment]
> > The fix provided by @NordicBeaver works perfectly and can bypass the issue.
> 
> Just to confirm, with that you mean a manual `npm install @prisma/client` before you first run `npx prisma generate`, correct?

Correct. Installing the client manually bypasses the issue.

**Comment by janpio** [comment]
I just noticed that at the end of the error message, it gives a quite explicit error message what is failing:
```
Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.2
```
Why is Prisma even running `npm install -D prisma`, when Prisma CLI is clearly already installed!?

**Comment by janpio** [comment]
Internal note:

This seems to be related to the Node-API library for Windows `query_engine-windows.dll.node`. When using the escape hatch to return to the binary engine, 
```
PRISMA_CLI_QUERY_ENGINE_TYPE=binary
PRISMA_CLIENT_ENGINE_TYPE=binary
```
(for example in your `.env` file), the error does _not_ occur like it does described above. Instead you get this one:

```
C:\Users\Jan\Documents\throwaway\9184>npx prisma generate
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma

> prisma@3.0.2 preinstall C:\Users\Jan\Documents\throwaway\9184\node_modules\prisma
> node scripts/preinstall-entry.js


> prisma@3.0.2 install C:\Users\Jan\Documents\throwaway\9184\node_modules\prisma
> node scripts/install-entry.js

npm notice save prisma is being moved from dependencies to devDependencies
npm WARN 9184@1.0.0 No description
npm WARN 9184@1.0.0 No repository field.

+ prisma@3.0.2
updated 1 package and audited 2 packages in 1.474s
found 0 vulnerabilities

npm WARN 9184@1.0.0 No description
npm WARN 9184@1.0.0 No repository field.

npm ERR! code ENOENT
npm ERR! syscall rename
npm ERR! path C:\Users\Jan\Documents\throwaway\9184\node_modules\@prisma\client
npm ERR! dest C:\Users\Jan\Documents\throwaway\9184\node_modules\@prisma\.client.DELETE
npm ERR! errno -4058
npm ERR! enoent ENOENT: no such file or directory, rename 'C:\Users\Jan\Documents\throwaway\9184\node_modules\@prisma\client' -> 'C:\Users\Jan\Documents\throwaway\9184\node_modules\@prisma\.client.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Jan\AppData\Roaming\npm-cache\_logs\2021-09-20T19_03_16_214Z-debug.log
Error: Command failed with exit code 4294963238: npm install @prisma/client@3.0.2
```

**Comment by janpio** [comment]
> > The script tries to install `@prisma/client` and `prisma` if they doesn't exists on your package.json file.
> 
> This is interesting, because prisma is already present in my package.json, but from what I see the script is trying to re-install prisma too (I guess so because the package getting disrupted is prisma itself).

I can confirm Prisma is _always_ trying to install `prisma` and `@prisma/client`, and lets npm/yarn deal with it already being installed:

https://github.com/prisma/prisma/blob/d75d71e6e31eefbc77f211466ab75a8dfd27b4b6/packages/sdk/src/predefinedGeneratorResolvers.ts#L108-L109

This is the reason why you are experiencing a failure in the `prisma` installation while already having `prisma` installed and actually using it right at that time.

Theory why this only now started failing: The Node-API engine file is "kept in use" (whatever that might mean in this context), while the binary engine file was not before.

**Comment by Jolg42** [comment]
Happened to another user in public Slack with `3.1.1`
https://prisma.slack.com/archives/CBFFY3066/p1632645964035500

**Comment by LewisTehMinerz** [comment]
🤔 I'm trying to think why that file might have a lock now but it didn't before. Is there anything that uses that file now during the generation of the client?

**Comment by janpio** [comment]
@LewisTehMinerz The Node-API `library` engine file was only introduced generally with 3.x - so this makes sense. The code behind that (internally known as `LibraryEngine`) is new and different from the old one (`BinaryEngine`). (In https://github.com/prisma/prisma/issues/9184#issuecomment-923204553 I describe how to switch back to that implementation, and then not have this error any more).

We will need to understand why there is a file lock on Windows on that file and then find a way to fix that. (And also create a test for CI that triggers this behavior along the way - so we can confirm the fix)

**Comment by janpio** [comment]
New minimal reproduction:

1. `git clone https://github.com/janpio/prisma-issue-9184`
2. `cd prisma-issue-9184`
3. `npm install`
4. `npx prisma generate`
5. Boom.

**Comment by janpio** [comment]
And here is an isolated reproduction of what is happening: https://github.com/janpio/prisma-9184-minimal-repro

If you are on Windows:

1. `git clone https://github.com/janpio/prisma-9184-minimal-repro`
2. `cd prisma-9184-minimal-repro`
3. `npm install`
4. `npx ts-node script.ts`

This shows that a `require` keeps a file open on Windows and then does not let any process, like `npm install`, delete that file.

**Comment by janpio** [comment]
This problem also affects usage of `npx prisma` without a `package.json`: https://github.com/prisma/prisma/issues/9517 That reproduction is even more minimal.

**Comment by juanbomfim22** [comment]
> Try to install prisma client with: `yarn add @prisma/client` later try again with: `yarn prisma migrate dev`

Thanks, it works. I think the problem occured because I didn't had `@prisma/client` and when first run `yarn prisma migrate dev` it automatically installed, but somehow failed in the fourth step

**Comment by MaxiSantos** [comment]
I had the same issues when trying to install a library (yarn add node-cron).. I tried @NordicBeaver solution (https://github.com/prisma/prisma/issues/9184#issuecomment-917396506) but just went to step 3 (yarn add @prisma/client). But I got the same result. Then I realized I had my server running so after I stoped it everything went to normal

**Comment by aqrln** [comment]
Current plan:

* Don't install the Prisma CLI when it is already present in `package.json`. This fixes the issue with npm 6.
* Yarn still removes the DLL library in Prisma CLI even when `@prisma/client` (or any other unrelated package) is installed. If we detect that `@prisma/client` is missing and the project is using Yarn, `prisma generate` will abort and ask the user to install `@prisma/client` manually, since that's the only safe option with Yarn.

**Comment by janpio** [comment]
Possible addition to the plan: Split out Prisma CLI installation completely from the "Prisma Client" is missing condition, as that can also be useful when Prisma Client is already there.

**Comment by pfried** [comment]
I want to note that I accidentially ran into the same issue while prisma studio was running in the background

**Comment by Jolg42** [comment]
Hi @pfried, if this still happens with the latest version of Prisma could you open a new issue with as many details you can? So we can try to reproduce.

**Comment by pfried** [comment]
@Jolg42 I do not think this is an error, because if you have a file in use by another process (the node dll) this is just the normal behaviour.

Just wanted to leave the comment because it could be helpful to others if they have the same issue

**Comment by LeonardoVini** [comment]
As @pfried mentioned, this only happened to me while I was running prisma studio in a background process.
I stopped prisma studio and installed the dependency I needed and it worked normally.
I don't know why this happens but it would be great if they fixed it

**Comment by joao-pedrozo** [comment]
I fixed by killing prisma studio process

**Comment by MrDrummer** [comment]
I stumbled across this just now. I do not have a `Prisma studio` running. I have tried with and without my nest.js app running in the background.

Deleting node_modules magically fixed it for me. I'll come back if I experience this again.

**Comment by adiozdaniel** [comment]
> Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> 
> 1. Delete `node_modules`
> 2. Run `npm install`
> 3. Run `npm install @prisma/client`
> 4. Run `npx prisma generate` / `npm prisma migrate dev`

This has just worked for me. Thanks

**Comment by adamsimonini** [comment]
> Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> 
> 1. Delete `node_modules`
> 2. Run `npm install`
> 3. Run `npm install @prisma/client`
> 4. Run `npx prisma generate` / `npm prisma migrate dev`

Shockingly this worked for me. To be clear I had to distinct, yet connected errors:

1. even as an admin, I wasn't able to delete a file generated by the prisma client within my node_modules folder
2. after painfully figuring out how to delete that folder, and doing an npm i, "npx prisma generate" failed due to a permissions issue.

Only after deleting the file, node_modules folder, re-installing from my package.json (which already includes prisma and @prisma/client), and then "npm i @prisma/client" explicitly after my npm -i, was I able to get "npx prisma generate" to work.

To be clear the offical prisma docs state the following: "Note also that prisma generate is automatically invoked when you're installing the @prisma/client npm package. So, when you're initially setting up Prisma Client, you can typically save the third step from the list above."

**Comment by oksimano** [comment]
Hy!

I have a same error but not the begining. 
I want install class-validator and class-transformer

After that i reinstall prisma and @prisma/client , and that i install class-validator and class-transformer.

**Comment by StrahinjaWebDev** [comment]
The problem with this can be if you try building projects etc npm run build, and you started your Prisma, etc npx Prisma studio. The project throws this error because it cannot build a project if Prisma is running.

**Comment by Jolg42** [comment]
Thank you both @StrahinjaWebDev & @ishaqyusuf for reporting this in comments here, I randomly found them.

Could you actually create a new bug issue, so we can look at it again with fresh eyes on a fresh version of Prisma? 

Note: creating a new bug issue (and mentioning this one) is the best way to have us look at it, comments in closed issues are very likely to be unseen.

**Comment by Mikask1** [comment]
Fixed the issue on my end. Turns out my server was still running and that was blocking access to specific files. Stopped the server, and did `npx prisma migrate dev` and it worked.

**Comment by jchewitt** [comment]
May help some, if your application is running while trying to migrate then prisma can't unlink. If that's the case just kill your node process and try again. I've done many times and forgotten that I still have my server running.

**Comment by dev03cyno** [comment]
i was running prisma studio in background kill your all  open terminals and then try

**Comment by elijaholmos** [comment]
getting the same error- I cannot `prisma:generate` if the client is already being used in an app which is running

**Comment by lumgashi** [comment]
I dont know if anyone is having this issue anymore, but what helped me in my case is that first I stopped the server, then ran npx commands and it worked perfectly fine.

**Comment by notflip** [comment]
> I dont know if anyone is having this issue anymore, but what helped me in my case is that first I stopped the server, then ran npx commands and it worked perfectly fine.

That's it! That's working for me, I'm using pm2, so after running `npx pm2 stop all` and re-running `npx prisma db push` it's working fine. Thanks @lumgashi

**Comment by JairoDosSantos** [comment]
Just remove node modules folder.

**Comment by edmblue** [comment]
> I dont know if anyone is having this issue anymore, but what helped me in my case is that first I stopped the server, then ran npx commands and it worked perfectly fine.

As simple as that, thanks

**Comment by taylorparrish2617** [comment]
For anyone still experiencing this issue, and using next.js... It would appear the next build can get corrupted sometimes. With the dev server not running, delete the .next build folder, then try npx prisma generate. Worked for me.

**Comment by douglasrcjames** [comment]
This bug popped up for me again, and seemed to coincide with me placing the first entry for a new model that I hadn't used yet.

**Comment by JuaniSilva** [comment]
same problem for me when I updated prisma

**Comment by bdthombre** [comment]
I go this problem, and I noticed that .dll file were running due to my current terminal was running "npm dev run"
I stopped the run and then the command was working nicely again. 
TRY STOPPING ACTIVE "npm run dev" or "npm start"

**Comment by Schmell** [comment]
> [lumgashi](/lumgashi)

I can see that this works, but honestly it is a pretty terrible DX.
Is anyone working on a fix for this??

**Comment by Schmell** [comment]
> I dont know if anyone is having this issue anymore, but what helped me in my case is that first I stopped the server, then ran npx commands and it worked perfectly fine.

I can see that this works, but honestly it is a pretty terrible DX.
Is anyone working on a fix for this??

**Comment by janpio** [comment]
Windows does not let us delete a file during the command you are trying to run. I don't think there is much we, Prisma, can do about that - or is there? Suggestions welcome.

**Comment by NatiG100** [comment]
I think prisma is trying to delete the file "query_engine-windows.dll.node" while at the same time the node.js environment is using it. So the easiest solution I found is to stop the server you are using and run the command that was previously causing this error.

**Comment by ricpham** [comment]
> Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> 
> 1. Delete `node_modules`
> 2. Run `npm install`
> 3. Run `npm install @prisma/client`
> 4. Run `npx prisma generate` / `npm prisma migrate dev`

Thanks, it work for me !

**Comment by SpinBoxx** [comment]
> I go this problem, and I noticed that .dll file were running due to my current terminal was running "npm dev run" I stopped the run and then the command was working nicely again. TRY STOPPING ACTIVE "npm run dev" or "npm start"

Yeah stopping the dev server worked for me

**Comment by Lasang3012** [comment]
In my case, i have to delete node_module and yarn install all package again, and i found reason lead to this error, because prisma use nodejs version > 14 or 15, but when i npx prisma generate to migrate i use nodejs version 14.17.3. It's lead to that error. ( because to run my project i must use nodejs version 14, and when i migrate i have to delete node_module )

To resolve that error, i found two ways to fix:
1) 
Tried re-installing @prisma/client manually, it seems to help. So here is a possible workaround until the problem is fixed.

Delete node_modules
Run npm install
Run npm install @prisma/client
Run npx prisma generate / npm prisma migrate dev

2) Use nodejs version >= 15 ( or in my case found what package use nodejs <15 and remove it to run completedly perfect project )

**Comment by filipjnc** [comment]
Stopping the dev server worked for me too, but is there really no way to circumvent this? It's an annoying thing to do.

**Comment by janpio** [comment]
Suggestions welcome @filipjnc, we see the problem but have no idea what to do about it.

**Comment by Lunatic02** [comment]
@filipjnc I did the same and it worked.

**Comment by Khamitamantaev** [comment]
Попробуйте остановить runtime приложение, которое у вас запущено в данный момент, после остановки приложения, yarn prisma generate

**Comment by joaquintalice** [comment]
If you're using typescript try this
`ctrl+p`
Write this
`>TypeScript: Restart TS Server`
That's all

**Comment by hemantkjalin** [comment]
Must stop server before running command.

**Comment by Schmell** [comment]
Prisma 5 has a `--no-engine` flag for this. As far as i can tell ( _without doing actual research_ :) ) this on works with `npx prisma generate`
Otherwise stop the dev server before generating as advised earlier in this issue

**Comment by Helios15** [comment]
I have error : "EPERM: operation not permitted, unlink 'C:\Users\ADMIN\Dfantasy\dFantasy-BE\node_modules\@prisma\engines\schema-engine-windows.exe", i tried to delete node_modules and manually reinstall prisma and @prisma/client but not work. Help me pls 

> Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> 
> 1. Delete `node_modules`
> 2. Run `npm install`
> 3. Run `npm install @prisma/client`
> 4. Run `npx prisma generate` / `npm prisma migrate dev`

I have error : "EPERM: operation not permitted, unlink 'C:\Users\ADMIN\Dfantasy\dFantasy-BE\node_modules\@prisma\engines\schema-engine-windows.exe", i tried to delete node_modules and manually reinstall prisma and @prisma/client but not work. Help me pls

**Comment by cfex** [comment]
> > Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> > 
> > 1. Delete `node_modules`
> > 2. Run `npm install`
> > 3. Run `npm install @prisma/client`
> > 4. Run `npx prisma generate` / `npm prisma migrate dev`
> 
> I have error : "EPERM: operation not permitted, unlink 'C:\Users\ADMIN\Dfantasy\dFantasy-BE\node_modules@prisma\engines\schema-engine-windows.exe", i tried to delete node_modules and manually reinstall prisma and @prisma/client but not work. Help me pls

You have to stop the server and then run `npx prisma generate`.

**Comment by mohankr123** [comment]
For IISNode:- 
1. stop site from application pool 
2. run command prisma generate
3. now start application pool

This works for me.

**Comment by satwikrprabhu** [comment]
yepp just kill the terminal, run prisma generate and npm run dev

**Comment by Kizito007** [comment]
> 4\. npm prisma migrate dev

Yes! I stopped the server, deleted my previous migration folder because I was trying to map, then ran `npx prisma migrate dev` and `npx prisma studio` and it worked.

**Comment by VicvekSr9485** [comment]
> 4\. npx prisma generate

This works for me, thanks

**Comment by samueloshio** [comment]
> Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> 
> 1. Delete `node_modules`
> 2. Run `npm install`
> 3. Run `npm install @prisma/client`
> 4. Run `npx prisma generate` / `npm prisma migrate dev`

This works for me. Thanks

**Comment by imf4isal** [comment]
Stopping the dev server has worked for me.

**Comment by kitsuneremi** [comment]
i received this error when tried to generate when the code still running on it

**Comment by aramrw** [comment]
No need to uninstall node or prisma you just need stop the dev server

**Comment by Ksahil-bazard** [comment]
> Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> 
> 1. Delete `node_modules`
> 2. Run `npm install`
> 3. Run `npm install @prisma/client`
> 4. Run `npx prisma generate` / `npm prisma migrate dev`

Can confirm faced this error and this method resolved it.

**Comment by cannibalhulk** [comment]
> Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> 
> 1. Delete `node_modules`
> 2. Run `npm install`
> 3. Run `npm install @prisma/client`
> 4. Run `npx prisma generate` / `npm prisma migrate dev`

Worked for me. But when I ran this `npx prisma db push` after that, it threw this error in terminal:

```
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": MongoDB database "nextauth" at "cluster0.pecwetf.mongodb.net"
Error: MongoDB error
Kind: SCRAM failure: bad auth : authentication failed, labels: {}
   0: schema_core::commands::schema_push::Calculate `from`
             at schema-engine\core\src\commands\schema_push.rs:29
   1: schema_core::state::SchemaPush
             at schema-engine\core\src\state.rs:436
```

**Comment by allgloryforrobots** [comment]
![image](https://github.com/prisma/prisma/assets/75021483/9d753548-48ea-43b6-bee0-b48aa82a74d3)

**Comment by allgloryforrobots** [comment]
2023 - Same bug

**Comment by shengzhou1216** [comment]
> Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> 
> 1. Delete `node_modules`
> 2. Run `npm install`
> 3. Run `npm install @prisma/client`
> 4. Run `npx prisma generate` / `npm prisma migrate dev`

Manually install `@prisma/client` , then `npm install` worked for me. It's wired 😥 .

**Comment by socertis** [comment]
all the above solutions didn't work for me, so, I searched for a hanging NodeJS process in my task manager, I found a process and killed it. and then, it is resolved.

**Comment by Shigbeard** [comment]
Just had this one, I found that the problem was that there was a nodejs instance still running that had imported prisma. Resolved by closing node. Recommend this be closed with a hotfix to better catch and label this error.

**Comment by openspaceranger** [comment]
NordicBeaver's fix was only temporary for me. I found that running powershell in administrator rights did the trick in my case. However, it's obviously not an ideal solution.

**Comment by justinstenning** [comment]
Just had this issue - in my case I think the cause was that I used a different NodeJS version since the last time I ran `npm install`, removing node_modules then `npm install` fixed the issue.

**Comment by BarunBlog** [comment]
Stop the dev server and run the `npm prisma migrate dev` command. It should work

**Comment by clemncls** [comment]
removing the node modules entirely worked for me

**Comment by aliarman89067** [comment]
if you are working with nextjs stop the server and then run npx prisma generate this worked for me :)

**Comment by abdulbasit0-UI** [comment]
stop your npm server before running prisma generate

**Comment by kshanu12** [comment]
> > > Can confirm, just had the same error. Tried installing `@prisma/client` manually, it seems to help. So here is a possible workaround until the problem is fixed.
> > > 
> > > 1. Delete `node_modules`
> > > 2. Run `npm install`
> > > 3. Run `npm install @prisma/client`
> > > 4. Run `npx prisma generate` / `npm prisma migrate dev`
> > 
> > 
> > I have error : "EPERM: operation not permitted, unlink 'C:\Users\ADMIN\Dfantasy\dFantasy-BE\node_modules@prisma\engines\schema-engine-windows.exe", i tried to delete node_modules and manually reinstall prisma and @prisma/client but not work. Help me pls
> 
> You have to stop the server and then run `npx prisma generate`.

This worked for me

**Comment by anuradhaindika83** [comment]
If there is another node instance running, then stop it and try again.

**Comment by akashjangra2002** [comment]
> If you're using typescript try this `ctrl+p` Write this `>TypeScript: Restart TS Server` That's all
> 
> (Remember stop server before running npx prisma generate)

I can confirm, this works for me too.

**Comment by Yasser-Alnajjar** [comment]
This usually happens because you are trying to do this and the development mode is working
Turn off development mode and try again. I think this will solve the problem

**Comment by yashrajt0001** [comment]
> ### Bug description
> Following the [Getting Started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) guide, when it gets to the point of running the "migrate dev" command, the script tries to install some packages through npm (don't know which ones) and this causes the error in title.
> 
> ### How to reproduce
> 1. Follow the [Getting started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) prisma guide up to and included the "npx prisma migrate dev --name init" command.
> 2. Get the following error:
> 
> ```
> Environment variables loaded from .env
> Prisma schema loaded from prisma\schema.prisma
> Datasource "db": PostgreSQL database "testprisma", schema "public" at "localhost:5432"
> 
> PostgreSQL database testprisma created at localhost:5432
> 
> The following migration(s) have been created and applied from new schema changes:
> 
> migrations/
>   └─ 20210909080511_init/
>     └─ migration.sql
> 
> Your database is now in sync with your schema.
> 
> Running generate... (Use --skip-generate to skip the generators)
> npm WARN hello-prisma@1.0.0 No description
> npm WARN hello-prisma@1.0.0 No repository field.
> 
> npm ERR! code EPERM
> npm ERR! syscall unlink
> npm ERR! path C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node
> npm ERR! errno -4048
> npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'
> npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
> npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
> npm ERR!     errno: -4048,
> npm ERR!     code: 'EPERM',
> npm ERR!     syscall: 'unlink',
> npm ERR!     path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node'
> npm ERR!   },
> npm ERR!   errno: -4048,
> npm ERR!   code: 'EPERM',
> npm ERR!   syscall: 'unlink',
> npm ERR!   path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node',
> npm ERR!   parent: 'hello-prisma'
> npm ERR! }
> npm ERR!
> npm ERR! The operation was rejected by your operating system.
> npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
> npm ERR! or that you lack permissions to access it.
> npm ERR!
> npm ERR! If you believe this might be a permissions issue, please double-check the
> npm ERR! permissions of the file and its containing directories, or try running
> npm ERR! the command again as root/Administrator.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     C:\Users\mbelletti\AppData\Roaming\npm-cache\_logs\2021-09-09T08_05_23_557Z-debug.log
> Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.1
> The batch file cannot be found.
> ```
> 
> The log file is the following: [2021-09-09T08_05_23_557Z-debug.log](https://github.com/prisma/prisma/files/7134677/2021-09-09T08_05_23_557Z-debug.log)
> 
> After this, the node_modules.bin directory no longer has the prisma executable, and the node_modules/prisma folder only contains the file that npm wasn't able to remove (query_engine-windows.dll.node).
> 
> ### Expected behavior
> The script finishes correctly.
> 
> ### Prisma information
> All the prisma configurations are the ones taken from the guide.
> 
> ### Environment & setup
> * OS: Windows 10
> * Database: PostgreSQL 13
> * Node.js version: 14.17.1 (using nvm)
> 
> ### Prisma Version
> ```
> Environment variables loaded from .env
> prisma                  : 3.0.1
> @prisma/client          : Not found
> Current platform        : windows
> Query Engine (Node-API) : libquery-engine 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\query_engine-windows.dll.node)
> Migration Engine        : migration-engine-cli 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\migration-engine-windows.exe)
> Introspection Engine    : introspection-core 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\introspection-engine-windows.exe)
> Format Binary           : prisma-fmt 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
> Default Engines Hash    : 2452cc6313d52b8b9a96999ac0e974d0aedf88db
> Studio                  : 0.423.0
> ```
> 
> (I had to relaunch npm install to get prisma back)

Problem: your node server is running somewhere and using the prisma client, even if you have closed the terminal sometimes it keeps running, thats why the error is saying to first stop the prisma client and then execute the command

Solution: Just restart your computer and it will be solved!

**Comment by willmagna** [comment]
Update the `@prisma/cliente`. And just delete the `query_engine-windows.dll.node` file and run `npx prisma generate` or `npx prisma migrate` again.

It worked for me

**Comment by ROHAN13498** [comment]
> Stop the dev server and run the `npx prisma migrate dev` command. It should work

This Worked Thanks!!🤝

**Comment by JimmyAG** [comment]
While deleting the `node_modules` folder, I found that  `query_engine-windows.dll.node` is open in node, and closing the dev server and prisma studio might also be a fix, instead of removing the `node_modules` folder

**Comment by sixtusagbo** [comment]
Try closing any development servers you have that uses the prisma client and then try running the command again. That's how I got mine to work.

**Comment by dorlanpabon** [comment]
> Al eliminar la `node_modules`carpeta, descubrí que `query_engine-windows.dll.node`está abierta en el nodo, y cerrar el servidor de desarrollo y prisma studio también podría ser una solución, en lugar de eliminar la `node_modules`carpeta.

Confirm solution, thanks you

**Comment by abderehmen02** [comment]
for me  , deleting node_modules and running npm i solves the issue .

**Comment by chiragyadav2003** [comment]
hey, you can try this : 
1. First try to delete node_modules/@prisma/client
2. Now run command - npx prisma generate
this will install new client

**Comment by noobpiyush** [comment]
> ### Bug description
> Following the [Getting Started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) guide, when it gets to the point of running the "migrate dev" command, the script tries to install some packages through npm (don't know which ones) and this causes the error in title.
> 
> ### How to reproduce
> 1. Follow the [Getting started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) prisma guide up to and included the "npx prisma migrate dev --name init" command.
> 2. Get the following error:
> 
> ```
> Environment variables loaded from .env
> Prisma schema loaded from prisma\schema.prisma
> Datasource "db": PostgreSQL database "testprisma", schema "public" at "localhost:5432"
> 
> PostgreSQL database testprisma created at localhost:5432
> 
> The following migration(s) have been created and applied from new schema changes:
> 
> migrations/
>   └─ 20210909080511_init/
>     └─ migration.sql
> 
> Your database is now in sync with your schema.
> 
> Running generate... (Use --skip-generate to skip the generators)
> npm WARN hello-prisma@1.0.0 No description
> npm WARN hello-prisma@1.0.0 No repository field.
> 
> npm ERR! code EPERM
> npm ERR! syscall unlink
> npm ERR! path C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node
> npm ERR! errno -4048
> npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'
> npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
> npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
> npm ERR!     errno: -4048,
> npm ERR!     code: 'EPERM',
> npm ERR!     syscall: 'unlink',
> npm ERR!     path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node'
> npm ERR!   },
> npm ERR!   errno: -4048,
> npm ERR!   code: 'EPERM',
> npm ERR!   syscall: 'unlink',
> npm ERR!   path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node',
> npm ERR!   parent: 'hello-prisma'
> npm ERR! }
> npm ERR!
> npm ERR! The operation was rejected by your operating system.
> npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
> npm ERR! or that you lack permissions to access it.
> npm ERR!
> npm ERR! If you believe this might be a permissions issue, please double-check the
> npm ERR! permissions of the file and its containing directories, or try running
> npm ERR! the command again as root/Administrator.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     C:\Users\mbelletti\AppData\Roaming\npm-cache\_logs\2021-09-09T08_05_23_557Z-debug.log
> Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.1
> The batch file cannot be found.
> ```
> 
> The log file is the following: [2021-09-09T08_05_23_557Z-debug.log](https://github.com/prisma/prisma/files/7134677/2021-09-09T08_05_23_557Z-debug.log)
> 
> After this, the node_modules.bin directory no longer has the prisma executable, and the node_modules/prisma folder only contains the file that npm wasn't able to remove (query_engine-windows.dll.node).
> 
> ### Expected behavior
> The script finishes correctly.
> 
> ### Prisma information
> All the prisma configurations are the ones taken from the guide.
> 
> ### Environment & setup
> * OS: Windows 10
> * Database: PostgreSQL 13
> * Node.js version: 14.17.1 (using nvm)
> 
> ### Prisma Version
> ```
> Environment variables loaded from .env
> prisma                  : 3.0.1
> @prisma/client          : Not found
> Current platform        : windows
> Query Engine (Node-API) : libquery-engine 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\query_engine-windows.dll.node)
> Migration Engine        : migration-engine-cli 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\migration-engine-windows.exe)
> Introspection Engine    : introspection-core 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\introspection-engine-windows.exe)
> Format Binary           : prisma-fmt 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
> Default Engines Hash    : 2452cc6313d52b8b9a96999ac0e974d0aedf88db
> Studio                  : 0.423.0
> ```
> 
> (I had to relaunch npm install to get prisma back)
If you are running nextjs server. Close the server and try running the command again.

$ npx prisma generate

**Comment by noobpiyush** [comment]
If you are running nextjs server. Close the server and try running the command again.

$ npx prisma generate

**Comment by khAja13** [comment]
Even I was facing this problem, then I have deleted the .prisma folder in node_modules, then I tried to generate and it was successful.

**Comment by wanderer-1994** [comment]
Notice that if I had my backend server running, I constantly got this error.
Once I stop the server, the error gone. My hunch is that the running server prevents deleting `query_engine-windows.dll.node`

**Comment by AliyanAamir** [comment]
Simple solution is stopping the server and then executing this command.

**Comment by thebjorn** [comment]
This is a prisma bug, stopping the server, restarting the pc, deleting the node_modules, are poor and inconvenient workarounds at best. For security reasons you cannot change an .exe file while you're running it, nor delete any .dll files it has loaded. This is for security and stability reasons. Prisma shouldn't do this.  There are windows APIs for doing this correctly and securely - prisma should use those APIs.

**Comment by Yaromey** [comment]
> Simple solution is stopping the server and then executing this command.

Maybe nice to tell the user that the the server is running and it could cause problems if you want to migrate? Looking trough a lot of stuff until I saw your comment, bit waste of time.

**Comment by janpio** [comment]
@thebjorn Can you tell us how? Optimally of course via a PR fixing this, but if you just explain it to us we would certainly consider changing the behavior so no one has to find this issue again in the future. Thanks!

**Comment by thebjorn** [comment]
Maybe this would work?: https://serverfault.com/a/503769/96844 In any case you'll need cooperation from the process holding the file open, otherwise you'll have to use the MOVEFILE_DELAY_UNTIL_REBOOT flag to MoveFileEx.

**Comment by marioalvesx** [comment]
> While deleting the `node_modules` folder, I found that `query_engine-windows.dll.node` is open in node, and closing the dev server and prisma studio might also be a fix, instead of removing the `node_modules` folder

This works for me, and after deleting it I just ran the migration command again and it's done. Been stuck on this error. Thanks, man!

OBS: Update the prisma version and prisma client version seens to be important too.
![image](https://github.com/prisma/prisma/assets/22736436/3d1a57ca-a8ac-42a3-9d05-c604a9644f7d)

**Comment by Somu878** [comment]
> It is because of the running next server. It is not able to mutate files as they are under usage, the OS is rejecting the request. CLose the server and editor then open the editor again and then try executing the same command. It will work. This worked for me.

This worked for me!

**Comment by AideFlood** [comment]
> This usually happens because you are trying to do this and the development mode is working Turn off development mode and try again. I think this will solve the problem

THANK YOU

**Comment by dev-phantom** [comment]
> 4\. npm prisma migrate dev

incase npm prisma migrate dev doesn't work here you can try
`npx prisma migrate dev` this is probablly because you don't have prisma installed globally 
you can install it globally and use `npm prisma migrate dev`

**Comment by sliterok** [comment]
Hey everyone, I found a workaround. You have to use the binary engine for development and kill "query-engine-windows.exe" process each time there's HMR event for prisma schema, then run prisma generate.
I implemented it as a vite plugin, you can check it out [here](https://www.npmjs.com/package/vite-plugin-prisma-hmr) [(code)](https://github.com/sliterok/vite-plugin-prisma-hmr/blob/master/src/index.ts).

**Comment by Adele583** [comment]
just run this script in your terminal, works like a charm:
# Identify the process using the file
$process = Get-Process | Where-Object {
    $_.Modules -match 'query_engine-windows.dll.node'
} | Select-Object Id, ProcessName, @{Name="Module";Expression={
    $_.Modules | Where-Object { $_.FileName -match 'query_engine-windows.dll.node' }
}}

# Display the process details (optional, for verification)
$process

# Terminate the process by its ID
if ($process) {
    Stop-Process -Id $process.Id -Force
    Write-Output "Process $($process.ProcessName) with PID $($process.Id) terminated."
} else {
    Write-Output "No process found using query_engine-windows.dll.node"
}

# Run prisma generate again after terminating the process
npx prisma generate

**Comment by clement-touzet** [comment]
> This usually happens because you are trying to do this and the development mode is working Turn off development mode and try again. I think this will solve the problem

Thank you it worked !

**Comment by Aestheticsuraj234** [comment]
Try to Shut Down Your Currently running server and then migrate

**Comment by 111akash** [comment]
> Try to Shut Down Your Currently running server and then migrate

Yep. This works. After turning off dev mode, it works fine.

**Comment by Ninh-Hung** [comment]
> Try to Shut Down Your Currently running server and then migrate

correctly!

**Comment by sharproyalz** [comment]
I have the same issue and I fixed it using these steps ps. it only works if you are using pnpm

- remove the prisma client in node modules.
 `pnpm remove @prisma/client` 

- add it back.
`pnpm add @prisma/client`

It will works now.

**Comment by ianlabicani** [comment]
> Try to Shut Down Your Currently running server and then migrate
Thank you so much for the suggestion to shut down the currently running server before migrating. It worked perfectly for me! I was having issues with migrations, and this simple step resolved the problem. I appreciate your help!

**Comment by omerman** [comment]
@janpio 
Im running the generate command as part of my build process.
It's not about "dev servers" I have production servers running, I won't just stop and re-start them.
It worked perfectly at prisma@4.11, which was the version I used just before upgrading now to 5.17.
I'm thinking I'll have to downgrade because of it.

Please help?
I wanted to upgrade not stay at the same place

**Comment by WhiteFoxCreamPie** [comment]
@omerman 

If you shutting down nextjs server, you can generate prisma with no error and restart nextjs server so now it working

**Comment by sachithrrra** [comment]
stop all instances of node.js running on your pc and run prisma generate

`taskkill /f /im node.exe`

**Comment by omerman** [comment]
I dont understand these comments. I dont want to stop all the instances of node.js..
If I stop them, my server will be down.

**Comment by WhiteFoxCreamPie** [comment]
@omerman
I know but this is the only way to reset this but don't worry you can restart your server after you generate prisma in just the few second

**Comment by omerman** [comment]
@QTNAandPie Im aware.. But I have multiple nextjs apps using prisma. I dont like the idea I have to do it.
Especially because I didn't have to do it in previous versions. It's a new issue. It should be fixed by Prisma. It's not a good solution to simply kill my prod nextjs instances in order to run a script that simply generates a typescript file 🤷 
I actually downgraded to Prisma v4.11 in order to keep working as usual..

**Comment by electrovir** [comment]
I ran into this while running tests on Windows CI runners. To ensure the tests reset their state, some of them delete `node_modules/.prisma` (due to what the tests are testing).

There is no server to kill, so I fixed this by calling `await prismaClient.$disconnect()` at the end of each test that constructed a `PrismaClient` instance.

**Comment by alvieridev** [comment]
> for me , deleting node_modules and running npm i solves the issue .

Yeah this worked for me too.

**Comment by omerman** [comment]
I 

> I ran into this while running tests on Windows CI runners. To ensure the tests reset their state, some of them delete `node_modules/.prisma` (due to what the tests are testing).
> 
> There is no server to kill, so I fixed this by calling `await prismaClient.$disconnect()` at the end of each test that constructed a `PrismaClient` instance.

I understand what you are saying.. but the thing is.. Im not running tests, Im not willing to disconnect prisma or kill the server.. I dont think there should be any locking of files while performing a basic operation such as generating types. I can't understand the need to lock a file while doing so 🤷

**Comment by heliang2502103** [comment]
> 如果您运行的是 nextjs 服务器。关闭服务器，然后再次尝试运行该命令。
> 
> $ npx prisma generate

感谢，你的建议解决了我的问题

**Comment by heliang2502103** [comment]
> 如果您运行的是 nextjs 服务器。关闭服务器，然后再次尝试运行该命令。
> 
> $ npx prisma generate

感谢，你的建议解决了我的问题

**Comment by iki** [comment]
**Workaround:** The following `generate` task in `package.json` works for me, when I have dev server with Prisma running:

```sh
rm -rf .prisma.tmp && mv -f ../node_modules/.prisma/client .prisma.tmp; prisma generate
```

That cleans up `.prisma.tmp` and moves Prisma client, which works OK on contrary to removing Prisma client directly, which fails with EPERM.

Note, that I'm in monorepo, so I have Prisma client in `../node_modules`. On single repo just use `node_modules` in current directory.

**Comment by omerman** [comment]
In dev mode it doesnt really matter. because you can simply stop your running code.

The issue here is in prod environments. I also have a monorepo that unfortunately runs on windows machine.
Prisma's engine file is locked as long as the process using it is running.(in windows) I dont think it is an actual "must" by Prisma.

I dont understand why this issue is not dealt with.

**Comment by iki** [comment]
@omerman It matters for me in dev mode. I have a monorepo with running dev servers for API (Prisma/Apollo) and Web (Next.js) and the Web generate task (GraphQL codegen) depends on API generate task (Nexus GraphQL codegen) in Turborepo orchestration, as well as the build tasks do.

So I need the above workaround to make `prisma generate` run smoothly even when dev server with Prisma is running. Works ok on Windows/Mac/Linux.

**Comment by ankur0999** [comment]
hey, i have the same problem 
this was happening because my backend was on in the another terminal
first close it then do npx prisma migrate dev/npx prisma generate

**Comment by JesusGMM** [comment]
Me funciono hacer lo que dice
@ankur0999 
hey, i was having the same problem
this was happening because my backend was on in the another terminal
first close it then do npx prisma migrate dev/npx prisma generate

**Comment by zeeshanalico** [comment]
All you have to do is just update the npm package using command
`npm install -g npm@latest`
and then delete node_modules and install again all the dependencies using 
`npm install`

**Comment by thoughtpalette** [comment]
Looks like mine was a permissions error. I needed to run as sudo or give the folders read/write perms.

**Comment by Rakibulbanna** [comment]
It worked when I terminated all open terminals and ran `npx prisma migrate dev --name <name> & npx prisma create.`

**Comment by dev-diva1** [comment]
> It is because of the running next server. It is not able to mutate files as they are under usage, the OS is rejecting the request. CLose the server and editor then open the editor again and then try executing the same command. It will work. This worked for me.

This is helped me. If you are using Nextjs ,close the server and then run the generate command.

**Comment by mustafazuberi** [comment]
![image](https://github.com/user-attachments/assets/45b82f3c-d543-4036-827d-739f078c7894)
If you're getting errors with npx prisma generate or migrate, close all running Prisma processes (e.g., Prisma Studio, Docker containers, or your app). Stop everything, then rerun the command, and it should work fine.

**Comment by ajaydewangan1100** [comment]
Close Conflicting Processes

Ensure no other processes (e.g., Node.js applications, editors like VS Code) are using the file or directory.
Close all active terminals or editors accessing the node_modules folder.

It worked for me

**Comment by BankaiTheOriginal** [comment]
I closed conflicting processes that might use the prisma service i.e (NestJS server) and ran 
`npx prisma generate` and it worked.

**Comment by Typeve** [comment]
I'm encountering the persistent `EPERM: operation not permitted` error when running `prisma generate` or `prisma db push` during active development on Windows. The error specifically blocks modifications to `query_engine-windows.dll.node` while the Node process is running (e.g., in watch mode or with a dev server).  

While stopping the Node process resolves the issue temporarily, this disrupts workflows like HMR (Hot Module Replacement) or persistent debugging sessions. Many developers in this thread have noted the same inconvenience.

**Questions/Suggestions:**  
1. Is there a way for Prisma to gracefully handle this file lock conflict? For example:  
   - Automatically terminating/restarting the engine process during generation.  
   - Using atomic file replacement (e.g., writing to a temp file first, then renaming).  
2. Are there alternative workarounds that don’t require killing the Node process?

**Comment by aqrln** [comment]
@Typeve 

> Automatically terminating/restarting the engine process during generation.

There's no separate engine process, the Node.js process _is_ the engine process, as the engine is loaded into the Node.js process as a native DLL.

> Using atomic file replacement (e.g., writing to a temp file first, then renaming).

Not sure it would help in this case but might be worth checking.

> Are there alternative workarounds that don’t require killing the Node process?

Using [WSL](https://learn.microsoft.com/en-us/windows/wsl/install) is a good option as Linux doesn't have these weird quirks.

**Comment by Nathanmedo** [comment]
I was having this issue just now but i just solved it, first of all close your editor, and stop the running nest server, stopping the server is important because it cannot migrate while the server is still running 
This was my error specifically:
`Your database is now in sync with your schema.

EPERM: operation not permitted, rename 'C:\Users\NATHAN\Desktop\Projects\nest js app\bookmark-app\generated\prisma\query_engine-windows.dll.nod
e.tmp18256' -> 'C:\Users\NATHAN\Desktop\Projects\nest js app\bookmark-app\generated\prisma\query_engine-windows.dll.node' i face this issue when prisma is trying to migrate to my database locally which is just a running conntainer, how to stop this problem`

And when it was fixed:
`Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": PostgreSQL database "collaboraid", schema "public" at "localhost:5342"

Already in sync, no schema change or pending migration was found.

✔ Generated Prisma Client (v6.6.0) to .\generated\prisma in 206ms`

**Comment by wardvisual** [comment]
A simple fix for this issue is to delete only the folder `node_modules/.prisma`, without needing to remove the entire node_modules directory. Then run:

1. `npm install`
2. `npm run db:generate`

**Comment by BrunoPierca** [comment]
> A simple fix for this issue is to delete only the folder `node_modules/.prisma`, without needing to remove the entire node_modules directory. Then run:
> 
> * `npm install`
> 
> And then db:generate

Thanks, this worked 1rst try

**Comment by guplem** [comment]
> A simple fix for this issue is to delete only the folder `node_modules/.prisma` 

Did work but wasn't trivial.
It said `Access to the path 'query_engine-windows.dll.node' is denied.` anyways.

But I managed to delete it using VSCode, which failed to move it to the recycle bin, and then it succeeded to completely erase it without moving it to the bin.

**Comment by Nijinsha** [comment]
Seems like it cannot lock the prisma/client. See if there are any instance of the app running. Close it before run the command.
